File: libs/service_layer/search/SearchEngine/src/main/kotlin/net/dankito/service/search/specific/FindAllEntriesHavingTheseTagsResult.java
package net.dankito.service.search.specific;


import net.dankito.deepthought.model.Entry;
import net.dankito.deepthought.model.Tag;
import net.dankito.service.search.util.CombinedLazyLoadingList;

import java.util.Collection;
import java.util.List;


public class FindAllEntriesHavingTheseTagsResult {

  protected Collection<Entry> entriesHavingFilteredTags;

  protected Collection<Tag> tagsOnEntriesContainingFilteredTags;
  protected List<Tag> tagsOnEntriesContainingFilteredTagsList = null;


  public FindAllEntriesHavingTheseTagsResult(Collection<Entry> entriesHavingFilteredTags, Collection<Tag> tagsOnEntriesContainingFilteredTags) {
    this.entriesHavingFilteredTags = entriesHavingFilteredTags;
    this.tagsOnEntriesContainingFilteredTags = tagsOnEntriesContainingFilteredTags;
  }

  public Collection<Entry> getEntriesHavingFilteredTags() {
    return entriesHavingFilteredTags;
  }

  public int getTagsOnEntriesContainingFilteredTagsCount() {
    return tagsOnEntriesContainingFilteredTags.size();
  }

  public Collection<Tag> getTagsOnEntriesContainingFilteredTags() {
    return tagsOnEntriesContainingFilteredTags;
  }

  public Tag getTagsOnEntriesContainingFilteredTagsAt(int index) {
    if(index < 0 || index >= getTagsOnEntriesContainingFilteredTagsCount())
      return null;

    if(tagsOnEntriesContainingFilteredTagsList == null) {
      if(tagsOnEntriesContainingFilteredTags instanceof List)
        tagsOnEntriesContainingFilteredTagsList = (List<Tag>)tagsOnEntriesContainingFilteredTags;
      else
        tagsOnEntriesContainingFilteredTagsList = new CombinedLazyLoadingList<Tag>(tagsOnEntriesContainingFilteredTags);
    }

    return tagsOnEntriesContainingFilteredTagsList.get(index);
  }
}

File: libs/service_layer/search/SearchEngine/src/main/kotlin/net/dankito/service/search/specific/FindAllEntriesHavingTheseTagsResult.kt
package net.dankito.service.search.specific


import net.dankito.deepthought.model.Entry
import net.dankito.deepthought.model.Tag
import net.dankito.service.search.util.CombinedLazyLoadingList


class FindAllEntriesHavingTheseTagsResult(entriesHavingFilteredTags: List<Entry>, tagsOnEntriesContainingFilteredTags: List<Tag>) {

    var entriesHavingFilteredTags: Collection<Entry>
        protected set

    var tagsOnEntriesContainingFilteredTags: Collection<Tag>
        protected set
    protected var tagsOnEntriesContainingFilteredTagsList: List<Tag>? = null


    init {
        this.entriesHavingFilteredTags = entriesHavingFilteredTags
        this.tagsOnEntriesContainingFilteredTags = tagsOnEntriesContainingFilteredTags
    }

    val tagsOnEntriesContainingFilteredTagsCount: Int
        get() = tagsOnEntriesContainingFilteredTags.size

    fun getTagsOnEntriesContainingFilteredTagsAt(index: Int): Tag? {
        if (index < 0 || index >= tagsOnEntriesContainingFilteredTagsCount)
            return null

        if (tagsOnEntriesContainingFilteredTagsList == null) {
            if (tagsOnEntriesContainingFilteredTags is List<*>)
                tagsOnEntriesContainingFilteredTagsList = tagsOnEntriesContainingFilteredTags as List<Tag>
            else
                tagsOnEntriesContainingFilteredTagsList = CombinedLazyLoadingList(tagsOnEntriesContainingFilteredTags)
        }

        return tagsOnEntriesContainingFilteredTagsList!![index]
    }
}
